fenics_helpers/timestepping.py

```diff
@@ -42,7 +42,32 @@ def __del__(self):
         if self._show_bar:
             self._pbar.close()
 
+class CheckPoints:
+    def __init__(self, points, t_start, t_end, dt_min=1e-6):
+        self.points = np.array(points)
+        self.post_process_t0 = False
+        if self.points.size != 0:
+            if self.points.max() > t_end or self.points.min() < t_start:
+                raise RuntimeError("Checkpoints outside of integration range.")
+            self.points = np.sort(np.unique(np.around(self.points, 6)))
+            if np.any(self.points < t_start + dt_min):
+                remaining = self.points >= t_start + dt_min
+                self.points = self.points[remaining]
+                self.post_process_t0 = True
+
+    def check(self, t, dt):
+        if self.points.size != 0 and self.points[0] < t + dt:
+            dt = self.points[0] - t
+            t = self.points[0]
```

Comment:
I was wondering for a long time, why we modify t here. Then I realized the `else` branch also modifies it. It's a small detail, but I would prefer the method to only return a modified `dt`. Maybe also rename to `timestep`. The call side would then look like
~~~
dt = checkpoints.timestep(t, dt)
t += dt
~~~

Then the actual advancement of the global time is still in the `TimeStepping`, not in the checkpoint. That way, restoring the previous time with `t -= dt` would be more clear.